pytensor/tensor/rewriting/math.py

```diff
@@ -423,6 +423,35 @@ def local_sumsqr2dot(fgraph, node):
                     return [new_out]
 
 
+@register_canonicalize
+@register_specialize
+@node_rewriter([Elemwise])
+def local_mulexp2expadd(fgraph, node):
+    """
+    This rewrite detects e^x * e^y and converts it to e^(x+y).
+    Similarly, e^x / e^y becomes e^(x-y).
+    """
+    if (
+        isinstance(node.op, Elemwise)
+        and isinstance(node.op.scalar_op, (aes.Mul, aes.TrueDiv))
+        and node.inputs[0].owner
+        and isinstance(node.inputs[0].owner.op, Elemwise)
+        and isinstance(node.inputs[0].owner.op.scalar_op, aes.Exp)
+        and node.inputs[1].owner
+        and isinstance(node.inputs[1].owner.op, Elemwise)
+        and isinstance(node.inputs[1].owner.op.scalar_op, aes.Exp)
+    ):
```

Comment:
`mul` can actually have arbitrary number of inputs such as in `mul(5, 6, 7)`, so the rewrite should be aware of that.

Reply:
The TrueDiv only accepts two inputs, never more, so you can keep treating as before.

You can assume multiplication is commutative yes. Careful that you don't want to sum (which reduces dimensions) but add all the cases that can be merged. Like `mul`, `add` allows arbitrary number of inputs.